### ISSUE TITLE
Screenshare has poor performance due to no hardware acceleration

### ISSUE THREAD

**Issue by MiMillieuh** [Vencord/Vesktop#191]
Hello, 

That's a native issue too but I may have a suggestion to make it work.

When you're streaming on discord (at least on wayland via xwayland or pipewire for now) the framerate of the app or game are extremely slow. But when discord is the focused window it's smooth again.

My suggestion would be to have a toggle (via a plugin or directly integrated) to keep the screenshare preview without having the window focused. (Disabled by default because I don't know it this issue affect windows and mac users. but it's not the case on the native client for windows at least.)

But that's only a suggestion and maybe that's not even what causes the issue... but it's weird that when the window is focused, this issue isn't happening.

**Comment by MiMillieuh** [comment]
> I've noticed only vesktop struggles with terrible screenshare performance (minutes per frame slow). Native discord only experiences random frame drops.

Well honestly I have the exact same issue on native client... 

Using OBS camera works fine tho

**Comment by maotovisk** [comment]
I'm also not able to stream at a decent frame rate, 60fps does not work and it does not even seem to be remotely close to 30.

**Comment by MrShelbySann** [comment]
Same here, I can't even stream in 30, it's like a slideshow. The issue is not present on the native client.

**Comment by Vendicated** [comment]
no. its on by default

**Comment by MiMillieuh** [comment]
I've tried to mess with hardware acceleration but it's still a slideshow...

**Comment by s-talay** [comment]
I think the native client for linux sucks because of electron not being able to utilize hardware acceleration? I don't know Electron all too well but maybe, since this project uses Gluon, there might be other ways of solving this issue?

**Comment by Vendicated** [comment]
we don't use gluon.

**Comment by ant0n-0x0000** [comment]
it can use hardware acceleration also

**Comment by s-talay** [comment]
> we don't use gluon.

I dont know how I possibly miss-read the readme this bad lol

**Comment by MiMillieuh** [comment]
Honestly I would like to try to fake having the discord window always on focus. 
I've noticed that when I click on my discord window and that the stream shows on the client (not showing the black thing to "use less ressources") it's way more smooth.

But is it possible to prevent that screen and function to come ?

**Comment by vickyorlo** [comment]
Are you sure that's really what fixes it? I've tried streaming things that don't require me to have the game focused (Beat Saber in particular) and the verdict was pretty clear - the framerate started dropping when there was a lot of movement on screen, but it would go back to relatively normal when there was not much happening. Tabbing back to Discord means you're not doing anything on the screen, which would improve bitrates. It seems to me like the upload is somehow getting throttled.

**Comment by MiMillieuh** [comment]
> Are you sure that's really what fixes it? I've tried streaming things that don't require me to have the game focused (Beat Saber in particular) and the verdict was pretty clear - the framerate started dropping when there was a lot of movement on screen, but it would go back to relatively normal when there was not much happening. Tabbing back to Discord means you're not doing anything on the screen, which would improve bitrates. It seems to me like the upload is somehow getting throttled.
> 
> Additionally, the local stream preview window is in no way reflective of the framerates other people watching the stream are getting - the preview can look absolutely fine while everybody else is getting maybe a frame every two seconds.

Well the only reason I think it's related is that when I'm on screenshare and I click on the discord window, everyone tells me that it's smooth now and right as I click away it's not anymore... 


So maybe it has something to to with that screen : 

![image](https://github.com/Vencord/Vesktop/assets/52078885/f06d896f-2db8-4c8f-93a8-9d0958095fc1)

**Comment by RinLovesYou** [comment]
Might be a long-standing [chromium bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016) that we could work around

**Comment by RinLovesYou** [comment]
It seems this was not, in fact, the issue.

**Comment by MiMillieuh** [comment]
Ok so I've tested today and it seems to be what's on screen, for some reasons Discord struggle to send it correctly as when there aren't that much movement it's smooth but when it's moving a lot it's a literal slide show (see video)

I also tried my theory about the focused window with this little code in the console : 

```
document.hasFocus = function() {
    return true;
};
```

But it seems that the preview is smooth even thought the other client receive a slide show... So it has nothing to do with it.

On the video most noticable things are when I start to move the OBS window fast it begins to be a slide show, In TM, when I stay still the ads and flags are smooth but once I move it's a slideshow, also the bush at the end, when I stop it's getting smooth.


On the left that's the original screen (Via OBS) and on the right it's the other discord client (Via Discord, what a friend would see.)

https://github.com/Vencord/Vesktop/assets/52078885/910c8e18-3f67-4dc9-9896-137de9239c93

**Comment by vickyorlo** [comment]
That is what I said... but thanks for testing and confirming, nonetheless.

**Comment by AssisrMatheus** [comment]
Is it really not possible to turn off HW acceleration?? Why is #229 closed then

**Comment by simvux** [comment]
I've been having the problem shown in the video above, basically regardless of which client I use for discord. 

The theory I have is that the video isn't being properly compressed. When someone screenshares from Windows and there's a lot of movement then the video quickly tends to become *heavily* compressed. Looking like 144p or even worse in darker footage. Meanwhile; even though the framerate sucks, the frames themselves seem to be of much higher quality regardless of movement on Linux. So; it's probably two different ways of which whichever layer is responsible compensates for tight bitrate constraints.

**Comment by MiMillieuh** [comment]
> I've been having the problem shown in the video above, basically regardless of which client I use for discord.
> 
> The theory I have is that the video isn't being properly compressed. When someone screenshares from Windows and there's a lot of movement then the video quickly tends to become _heavily_ compressed. Looking like 144p or even worse in darker footage. Meanwhile; even though the framerate sucks, the frames themselves seem to be of much higher quality regardless of movement on Linux. So; it's probably two different ways of which whichever layer is responsible compensates for tight bitrate constraints.

I've thought about it but also on windows the quality doesn't degrade that much depending on the connection. Also I've tested it and the streamer (on Linux Vestkop) sends about 0.5MB/S when there is a lot of movement, and about 0.1 - 0.2 MB/S when still.

Honestily I don't think 0.5MB/S is big but maybe we can't send more and we're sending more... IDK.

**Comment by oatmealine** [comment]
> I've been having the problem shown in the video above, basically regardless of which client I use for discord.
> 
> The theory I have is that the video isn't being properly compressed. When someone screenshares from Windows and there's a lot of movement then the video quickly tends to become _heavily_ compressed. Looking like 144p or even worse in darker footage. Meanwhile; even though the framerate sucks, the frames themselves seem to be of much higher quality regardless of movement on Linux. So; it's probably two different ways of which whichever layer is responsible compensates for tight bitrate constraints.

I've constantly had the regular Discord client downgrade my stream when it understands my internet can't handle it - this is okay, as I don't get very high speeds at all - but Vesktop never does this, instead opting to support its quality at all times.
Additionally, I don't actually know if the quality settings work? At least, I've never seen them reflected in the stream preview, when regular Discord always shows what's being sent

**Comment by MiMillieuh** [comment]
> > I've been having the problem shown in the video above, basically regardless of which client I use for discord.
> > 
> > The theory I have is that the video isn't being properly compressed. When someone screenshares from Windows and there's a lot of movement then the video quickly tends to become _heavily_ compressed. Looking like 144p or even worse in darker footage. Meanwhile; even though the framerate sucks, the frames themselves seem to be of much higher quality regardless of movement on Linux. So; it's probably two different ways of which whichever layer is responsible compensates for tight bitrate constraints.
> 
> I've constantly had the regular Discord client downgrade my stream when it understands my internet can't handle it - this is okay, as I don't get very high speeds at all - but Vesktop never does this, instead opting to support its quality at all times.
> Additionally, I don't actually know if the quality settings work? At least, I've never seen them reflected in the stream preview, when regular Discord always shows what's being sent

Previews are in both clients smooth, but I can definitely tell that the same low framerate issues happen to me on both clients when looking at what someone watching the stream would see... As for the quality, I can't really tell... That might even be another issue then.

**Comment by heliguy4599** [comment]
I wanted to join the discussion to say that I also indeed have this problem. My friends say that my streams are very noticeably bad, nearly unwatchable. I'm using an eGPU with my laptop and ran Vesktop upon the eGPU, still no dice. It seems the problem is not nearly as bad when running on my desktop.

**Comment by cyber-sushi** [comment]
I'll add my experience as well.
I've tried streaming using Vesktop on both my desktop PC (Nvidia GTX 1060) and my laptop (AMD Vega 8), and they run equally slow at "about 5 fps" according to my friends.

However, I've tried other solutions before, like Discord-Screenaudio, and they seem to suffer from the same issue, as opposed to the regular Discord client.
As others have mentioned, even though the framerate is low, the picture quality remains perfect, it never shows compression or artifacts which usually appear in fast scenes on the regular Discord client or on Windows.

I'm using Hyprland on a relatively weak internet connection (~3 Mbps upload).

**Comment by queenbiscuit311** [comment]
I've tried discord-screenaudio, vesktop, and pipewire-screenaudio all with the same results. Depending on the situation some clients can be better than others, but there's some apps that are so bad that in order to stream them I had to connect to my computer through moonlight and stream that on a windows or mac system. It seems to be the more multi core cpu usage the worse the stream, and i also cant seem to find any solid indication that discord streaming is actually using my video encoder.

**Comment by alexankitty** [comment]
I don't really know the code base well enough to definitively say what my findings mean, but looking through the console while having a stream going shows a couple of interesting things.
```
[UnifiedConnection(default)] updateVideoQuality: {
    "remoteSinkWantsMaxFramerate": 30
}

[UnifiedConnection(default)] BaseWebRTCConnection.updateVideoQuality: resolutionScaleFactor: undefined, encodingVideoWidth: undefined, remoteSinkWantsMaxFramerate: 30, encodingVideoMaxBitRate: undefined, localWant: 100

[UnifiedConnection(default)] UnifiedConnection.setTransceiverEncodingParameters: maxFramerate: from '20' to '30'
```
My best guess is it seems like the discord end is saying to change the framerate a few times... and I usually see it change between 20 to 30, but then the other issue I'm seeing is that the encodingVideoMaxBitRate becomes undefined, which I think might possibly be resulting in dropped frames? 
Looking at btop on my end, while there's a lot on the screen I don't really see the bitrate exceed 3.5mbps, while the maxbitrate is set to 8mbps and the target being 6mbps.
Perhaps there is something else discord needs to be told when building the stream?

There is also another oddity I'm seeing where the video quality suggest that the values provided by the application and user don't seem to be respected. For reference I'm streaming a 2560x1440 screen at 1080p60, but it appears the stream parameters get dropped to 720p30 at 2.5mbps max bitrate.
```
[UnifiedConnection(default)] updateVideoQuality: {
    "remoteSinkWantsPixelCount": 921600,
    "remoteSinkWantsMaxFramerate": 30,
    "encodingVideoMinBitRate": 150000,
    "encodingVideoMaxBitRate": 2500000,
    "encodingVideoBitRate": 2500000,
    "streamParameters": [
        {
            "type": "video",
            "active": false,
            "rid": "100",
            "ssrc": 101591,
            "rtxSsrc": 101592,
            "quality": 100,
            "maxBitrate": 2500000,
            "maxFrameRate": 30,
            "maxResolution": {
                "type": "fixed",
                "width": 1280,
                "height": 720
            },
            "maxPixelCount": 921600
        }
    ]
 ```

**Comment by splicedbread** [comment]
> Might be a long-standing [chromium bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016) that we could work around

> Edit: Web discord is also limited to 720p, a good portion of this might be a limitation of Vesktop technically being the browser version of discord.

From above, what made RinLovesYou to change their mind about that possible chromium bug? It looks like to me that when contentHint="detail" is passed with VP9 that it results in what MiMillieuh had found. 

Quality frames are being sent, possibly with the lowest 'default framerate values for quality. The receiving clients see high resolution non-compressed frames, as described in the issue and video evidence earlier that MiMillieuh provided.

Toggling between "detail" and "motion" options may result in a more fluid stream at 720p60fps with a lot more visible compression. 

I would like for this to be wrong, because the discussion on the chromium bug seems to have died out, when it should be discussed further over there : /

I am going to continue digging around...

**Comment by Plarpoon** [comment]
There is definitely something wrong tho, I have been using vesktop across different distros and physical machines for the past six months: A desktop, two laptops, a steam deck and distros like arch, pop-os, fedora, NixOS and now even Ubuntu. In all of these the screen share works fine but the frame rate is awful, often described to me as one frame every second. Apparently it would get better when I was not moving within the game, but as soon as it has to draw several things it goes down to nothing.

**Comment by Kagukara** [comment]
Thought this would be the best place to add this, using discord in librewolf/firefox on swaywm with the steps [here](https://wiki.archlinux.org/title/Firefox#Hardware_video_acceleration) to enable hardware acceleration (not sure if this has to be done), doesn't cause the screenshare to be a slideshow. Playing more GPU heavy games (GW2, Armored Core 6 etc) caused Electron/Chromium clients screenshare to be a slideshow. But when I screenshared Risk of Rain Return it was smooth.

Would it be possible to switch to the client to being firefox based or is that not possible?

Having the resolution/framerate/audio select is really nice, but isn't available when using the userscript to add vencord to librewolf/firefox.

**Comment by chamychu** [comment]
> Thought this would be the best place to add this, using discord in librewolf/firefox on swaywm with the steps [here](https://wiki.archlinux.org/title/Firefox#Hardware_video_acceleration) to enable hardware acceleration (not sure if this has to be done), doesn't cause the screenshare to be a slideshow. Playing more GPU heavy games (GW2, Armored Core 6 etc) caused Electron/Chromium clients screenshare to be a slideshow. But when I screenshared Risk of Rain Return it was smooth.
> 
> 
> 
> Would it be possible to switch to the client to being firefox based or is that not possible?
> 
> 
> 
> Having the resolution/framerate/audio select is really nice, but isn't available when using the userscript to add vencord to librewolf/firefox.

I think Firefox should be the move here 🙌

**Comment by Vendicated** [comment]
electron only supports chromium

**Comment by Plarpoon** [comment]
@Vendicated yeah, that's unfortunately very true. Do you think is there any way to use to leverage the VAAPI codecs already present in FlatHub? If you tap into those builds you should be able to have a fast video rendering with HW acceleration no? To give an idea this package has the full support for hardware rendering.

org.freedesktop.Platform.ffmpeg-full

**Comment by Kagukara** [comment]
> Thought this would be the best place to add this, using discord in librewolf/firefox on swaywm with the steps [here](https://wiki.archlinux.org/title/Firefox#Hardware_video_acceleration) to enable hardware acceleration (not sure if this has to be done), doesn't cause the screenshare to be a slideshow. Playing more GPU heavy games (GW2, Armored Core 6 etc) caused Electron/Chromium clients screenshare to be a slideshow. But when I screenshared Risk of Rain Return it was smooth.

I'm playing on a 4k120Hz display for more context.

Retested with different methods in Armored Core 6 Arena.
1. Electron/Chromium discord client (Used Vesktop) screenshare
   - Slideshow
   - Bitrate seemed fine
2. Librewolf discord browser screenshare
   - Better than a slideshow but seemed like 15fps
   - Bitrate seemed fine
3. Librewolf discord browser used camera (obs virtual camera)
   - Was like 30fps
   - Bitrate worse than 1&2
4. Electron/Chromium discord client (Used Vesktop) used camera (obs virtual camera)
   - Was laggy at the start but caught but and was like 30fps
   - Bitrate worse than 1&2

**Comment by Calosis** [comment]
> @Vendicated yeah, that's unfortunately very true. Do you think is there any way to leverage the VAAPI codecs already present in FlatHub? If you tap into those builds you should be able to have a fast video rendering with HW acceleration no? To give an idea this package has the full support for hardware rendering.
> 
> org.freedesktop.Platform.ffmpeg-full

This would be the best way to fix screen sharing using any discord client on Linux until Discord fixes it themselves.

Someone was working on proper hardware-accelerated screen sharing using FFMPEG as well as audio, but they seem to have given up on it. https://github.com/itsfolf/Discord-Screenshare-Linux/tree/hardware-encoding

**Comment by MiMillieuh** [comment]
> > @Vendicated yeah, that's unfortunately very true. Do you think is there any way to leverage the VAAPI codecs already present in FlatHub? If you tap into those builds you should be able to have a fast video rendering with HW acceleration no? To give an idea this package has the full support for hardware rendering.
> > org.freedesktop.Platform.ffmpeg-full
> 
> This would be the best way to fix screen sharing using any discord client on Linux until Discord fixes it themselves.
> 
> Someone was working on proper hardware-accelerated screen sharing using FFMPEG as well as audio, but they seem to have given up on it. https://github.com/itsfolf/Discord-Screenshare-Linux/tree/hardware-encoding

Honestly it's a really good project but it's not maintained anymore and it crashes a lot the whole VM or discord. and need to be reinstalled each update. Also audio screen share is hit or miss with this one.

**Comment by SpookySkeletons** [comment]
This issue is universal to wayland on multiGPU systems I've found, everything is affected. OBS, all screen recording applications, any share of windows. If it exists across two GPUs and requires encoding to be done you will suffer.

Not sure of the cause but typically linear memory formats are utilized to blit an image across multiple GPUs however something even stranger is going on here as it comes with a performance penalty but should NOT be this slow, this is a HARD choke somewhere on the encoder?? The more complex your bitrate requirements the more likely it is to choke.

Slight mitigation appears to be to mind which GPU performs the encoding but it's not a complete fix. Hosting encode on the same GPU doing the rendering looks to mitigate the worst of it.

Chromium does not let you choose which GPU to encode on or even anything but the CPU, most times.

This should probably be brought up with mesa or kernel devs to have a better look.

**Comment by t3rminus** [comment]
> This issue is universal to wayland on multiGPU systems I've found

it’s definitely not a multi-GPU issue. I am experiencing this issue with a single GPU (Radeon RX5700, if it matters). My CPU is a Ryzen 5 3600 that does not include a GPU.

I am using Wayland (Gnome), however, so the cause/solution might be there.

**Comment by AAGaming00** [comment]
Confirming not a Multi-GPU issue, my laptop has two GPUs and I do not have this issue.

**Comment by Plarpoon** [comment]
Confirming also not a gpu issue, strictly using a discrete GPU here

**Comment by rowbawts** [comment]
I don't think its ever been a multi GPU issue, OBS works perfectly for me (5800X has no iGPU & 6950XT). It's clearly just a lack of proper hardware acceleration for encoding the stream.

**Comment by yoyossef** [comment]
Not a Wayland issue either, happens on gamescope which uses X11 (Steam Deck) too

**Comment by MiMillieuh** [comment]
> I don't think its ever been a multi GPU issue, OBS works perfectly for me (5800X has no iGPU & 6950XT). It's clearly just a lack of proper hardware acceleration for encoding the stream.

I can confirm it's nothing related to since on the demo video I've provided you can see that OBS works perfectly fine

**Comment by Vendicated** [comment]
please stop leaving unproductive replies, you're just spamming both this issue thread and everyone's inbox. i don't wanna have to lock this issue

refrain from leaving further comments unless they are actually productive, for example if you found a relevant chrome bug thread, or if you found a solution

**Comment by SpookySkeletons** [comment]
Notable update: https://chromium-review.googlesource.com/c/chromium/src/+/3646633

**Comment by ve3xone** [comment]
We wrote about VAAPI above, I personally got it working with the following parameters: 
`vencord-desktop --enable-features=VaapiIgnoreDriverChecks,VaapiVideoEncoder,VaapiVideoDecoder,CanvasOopRasterization,UseMultiPlaneFormatForHardwareVideo` 

after such settings, the encode from the video card finally started working. 

My GPU: AMD Radeon RX 580 8gb.

Thank you!

**Comment by MiMillieuh** [comment]
> @SpookySkeletons thank you for the information about VAAPI, I personally got it working with the following parameters: `vencord-desktop --enable-features=VaapiIgnoreDriverChecks,VaapiVideoEncoder,VaapiVideoDecoder,CanvasOopRasterization,UseMultiPlaneFormatForHardwareVideo`
> 
> after such settings, the encode from the video card finally started working.
> 
> My GPU: AMD Radeon RX 580 8gb.

This arguments on my hardware aren't doing much, I guess that's maybe due to my optimus configuration. even if it use my intel GPU to encode.

Can you confirm if you still encounter 1 fps on another account while watching your stream ?

By the way I've noticed that when the stream becomes 1fps it shows this in the output when launched in a terminal : 

```
[OpenH264] this = 0x0x3c18093ae760, Warning:[Rc] iDid = 0,iContinualSkipFrames(3) is large
[OpenH264] this = 0x0x3c18093ae760, Warning:[Rc] iDid = 0,iContinualSkipFrames(3) is large
[OpenH264] this = 0x0x3c18093ae760, Warning:[Rc] iDid = 0,iContinualSkipFrames(3) is large
[OpenH264] this = 0x0x3c18093ae760, Warning:[Rc] iDid = 0,iContinualSkipFrames(6) is large
[OpenH264] this = 0x0x3c18093ae760, Warning:[Rc] iDid = 0,iContinualSkipFrames(3) is large
[OpenH264] this = 0x0x3c18093ae760, Warning:[Rc] iDid = 0,iContinualSkipFrames(3) is large
[OpenH264] this = 0x0x3c18093ae760, Warning:[Rc] iDid = 0,iContinualSkipFrames(3) is large
[OpenH264] this = 0x0x3c18093ae760, Warning:[Rc] iDid = 0,iContinualSkipFrames(3) is large
[OpenH264] this = 0x0x3c18093ae760, Warning:[Rc] iDid = 0,iContinualSkipFrames(3) is large

```

And when I stop moving a window fast, it doesn't output anything else.

**Comment by chamychu** [comment]
> @SpookySkeletons thank you for the information about VAAPI, I personally got it working with the following parameters: `vencord-desktop --enable-features=VaapiIgnoreDriverChecks,VaapiVideoEncoder,VaapiVideoDecoder,CanvasOopRasterization,UseMultiPlaneFormatForHardwareVideo` 
> 
> 
> 
> after such settings, the encode from the video card finally started working. 
> 
> 
> 
> My GPU: AMD Radeon RX 580 8gb.

Any results??

**Comment by Clifforus** [comment]
> @SpookySkeletons thank you for the information about VAAPI, I personally got it working with the following parameters: `vencord-desktop --enable-features=VaapiIgnoreDriverChecks,VaapiVideoEncoder,VaapiVideoDecoder,CanvasOopRasterization,UseMultiPlaneFormatForHardwareVideo`
> 
> after such settings, the encode from the video card finally started working.
> 
> My GPU: AMD Radeon RX 580 8gb.

Arch Linux, KDE, Wayland, RX 6700XT here.

Can confirm that without these options, my friends were saying my screen share was running at about 5-10FPS, but now I've tried with these options and they're saying it's smooth.

**Comment by SpookySkeletons** [comment]
Patched & confirming it's better than 1 fps, however watchers report about 5.
What's likely going on is I am now encoding on my iGPU instead of dGPU as expected which is unable to handle the image bitrate like @MiMillieuh 

https://wiki.archlinux.org/title/Firefox#Hardware_video_acceleration

Arch wiki reports that firefox can self select render node with the env var MOZ_DRM_DEVICE, what we could use is a chrome var to set the VAAPI dev here as well, that should mop up the issue for everyone involved.

Would anyone be willing to peek at Chromium's render node selection to slip in a patch here short term for PRIME users and try to get one upstream as well?

**Comment by Kagukara** [comment]
Using `vencord-desktop --enable-features=VaapiIgnoreDriverChecks,VaapiVideoEncoder,VaapiVideoDecoder,CanvasOopRasterization,UseMultiPlaneFormatForHardwareVideo` while playing Armored Core IV, it's a semi slideshow like 5fps. Arch Sway 7900XTX. Although `ENC` appears in `nvtop` so it looks like its encoding.

**Comment by Clifforus** [comment]
Upon further testing, it seems using `vesktop --enable-features=VaapiIgnoreDriverChecks,VaapiVideoEncoder,VaapiVideoDecoder,CanvasOopRasterization,UseMultiPlaneFormatForHardwareVideo` is definitely doing _something_ that's improving the overall stream performance, but the slowdowns when there's a lot of movement on screen persist, they're just not as extreme, and it also seems like the threshold for movement causing a slowdown is a bit higher. It also appears that lowering the stream resolution and the game resolution both help improve performance.

**Comment by SpookySkeletons** [comment]
Do list specs with your reports, I would like to narrow down if it's related to wrong GPU encoder selection.

Are you two on PRIME/ Hybrid GPU?

**Comment by Clifforus** [comment]
I gave some specs in a post above but I'll go into more detail here.

GPU: AMD Radeon RX 6700XT
CPU: Ryzen 7 5700X
OS: Arch Linux
DE: Plasma 5.27.10
WM: kwin
Display Server: Wayland
Audio Server: Pipewire
Kernel: 6.6.10-zen1-1-zen
Mesa version: 23.3.2-2
LLVM version: 16.0.6-1

The 6700XT is the only GPU in my system, so no PRIME.

**Comment by I-want-copilot-access-please** [comment]
I compiled Mesa with the PR [!26165,](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/26165). With my 7900XTX, I do see the "ENC" on nvtop, it goes from 0 to 4% while streaming, but it's still seems to lag a lot

**Comment by lou-perret** [comment]
I compiled Mesa with the PR [!26165,](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/26165). With my 7900XTX, I do see the "ENC" on nvtop, it goes from 0 to 4% while streaming, but it's still seems to lag a lot

**Comment by NextWork123** [comment]
we are sure the actual electron 28.1.1 have that commit merged? i'm not sure about that we need too the new electron with that merge i think

**Comment by notpeelz** [comment]
You can test it with chromium snapshot builds ([available since 122.0.6238.2](https://chromiumdash.appspot.com/commit/1ce9abc81fac01b1a1d91bddf57f37747fe5ced3)).

On Arch Linux, the most convenient way is to use the `chromium-snapshot-bin` AUR package.

**Comment by rurigk** [comment]
I installed Electron 29 (beta)
And added the flags from
https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/gpu/vaapi.md#verify-driver
Just to test 

```
app.commandLine.appendSwitch("enable-features", "VaapiVideoDecodeLinuxGL,VaapiVideoEncoder");
app.commandLine.appendSwitch("use-gl", "angle");
app.commandLine.appendSwitch("use-angle", "gl");
app.commandLine.appendSwitch("ozone-platform", "x11");
```

I can see in nvtop that its encoding with gpu and the frames are a lot smoother (still droping frames but not a slideshow anymore)

But i noticed something else, even if i select 420p as resolution the stream looks as crisp as my 1440p monitor if dont move in game, it is the resolution selection working?

**Comment by Krupski19** [comment]
> I installed Electron 29 (beta) And added the flags from https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/gpu/vaapi.md#verify-driver Just to test
> 
> ```
> app.commandLine.appendSwitch("enable-features", "VaapiVideoDecodeLinuxGL,VaapiVideoEncoder");
> app.commandLine.appendSwitch("use-gl", "angle");
> app.commandLine.appendSwitch("use-angle", "gl");
> app.commandLine.appendSwitch("ozone-platform", "x11");
> ```
> 
> I can see in nvtop that its encoding with gpu and the frames are a lot smoother (still droping frames but not a slideshow anymore)
> 
> But i noticed something else, even if i select 420p as resolution the stream looks as crisp as my 1440p monitor if dont move in game, it is the resolution selection working?

I can confirm this behaviour as well.  The quality of the image on the viewer's side does not seem to change at all when streaming at 480p or 1440p, the image at 480p looks near identical to the 1440p image, even though the stated stream resolution and frame rate reported to the viewer are corrected to the stream values selected.

One thing I would also like to point out is the behaviour of the encoder while streaming when using the VaapiEncoder, though I believe the same behaviour is occuring with CPU encoding, it's now just easer to visualize it.

When streaming a mostly static image or scence, the encoder utilization is somewhere between 10%-15%, regardless of the resolution or framerate settings selected.  Once there is more motion on the screen, encoder utilization drops to about 3%-5%, and the video turns into a slideshow.

**Comment by rejedai** [comment]
I noticed that the problem with video lags is not present when using discord-screenaudio, maybe it will help to figure out why this happens.

**Comment by Plarpoon** [comment]
The issue is not present at all in Fedora with the official Discord client, my viewers reported speeds and quality on par with the Windows counterpart. As far as I know the big difference is that Discord on Fedora uses Wayland to X11 video bridge under the hood.

**Comment by duckyondiscord** [comment]
> The issue is not present at all in Fedora with the official Discord client, my viewers reported speeds and quality on par with the Windows counterpart. As far as I know the big difference is that Discord on Fedora uses Wayland to X11 video bridge under the hood.

I think that's because the official client can do GPU acceleration, at least it has an option for it in its settings.

**Comment by Vendicated** [comment]
Issue unlocked. As mentioned above, please do not comment unless you have something valuable to add. Or in other words, do not comment unless you absolutely know what you're doing, and are either trying the command line flags mentioned above, or are (successfully) trying some other chromium/electron feature

___

Vesktop now officially uses electron 29

Could you please retry the vaapi chromium command line flags and see if they work for hardware accelerated screenshare on this electron version?

If yes, does this fix the issue? Does it reach 60fps or is it only marginally better, if at all?

**Comment by Kagukara** [comment]
Running Vencord with:
```
vesktop --enable-features=UseOzonePlatform --ozone-platform=wayland --use-gl=angle --use-angle=gl --enable-features=VaapiVideoDecodeLinuxGL,VaapiVideoEncoder
```
and changing my 4K display to 8 bit, the screenshare is still like a slideshow for others and smooth in the preview for me. The GPU is using more power when screensharing, but is not showing any indication like ENC in `nvtop`.

When screensharing, no matter the resolution I pick it looks like the same resolution. 480p60fps looks the same as 1440p60fps. Tested by watching the screenshare from an alternative account.

<details> <summary>Terminal output:</summary>

```
$ vesktop --enable-features=UseOzonePlatform --ozone-platform=wayland --use-gl=angle --use-angle=gl --enable-features=VaapiVideoDecodeLinuxGL,VaapiVideoEncoder
[181198:0312/032740.207314:ERROR:egl_dmabuf.cc(608)] Failed to record frame: Error creating EGLImage - EGL_BAD_MATCH
[181198:0312/032740.207336:ERROR:shared_screencast_stream.cc(940)] Dropping DMA-BUF modifier: 144115188757659396 and trying to renegotiate stream parameters
'loop->recurse > 0' failed at ../pipewire/src/pipewire/thread-loop.c:426 pw_thread_loop_wait()
[2024-03-12 03:27:41.053] [venmic] [info] [patchbay] (get) running venmic 3.3.2
[2024-03-12 03:27:41.058] [venmic] [info] [patchbay] (add_global) found metadata: 36
[181198:0312/032749.090299:ERROR:egl_dmabuf.cc(608)] Failed to record frame: Error creating EGLImage - EGL_BAD_MATCH
[181198:0312/032749.090322:ERROR:shared_screencast_stream.cc(940)] Dropping DMA-BUF modifier: 144115188757659396 and trying to renegotiate stream parameters
[181244:0312/032813.960901:ERROR:vaapi_video_decoder.cc(1242)] failed Initialize()ing the frame pool
[181198:0312/033103.759833:ERROR:egl_dmabuf.cc(608)] Failed to record frame: Error creating EGLImage - EGL_BAD_MATCH
[181198:0312/033103.759852:ERROR:shared_screencast_stream.cc(940)] Dropping DMA-BUF modifier: 144115188757659396 and trying to renegotiate stream parameters
[181198:0312/033104.736945:ERROR:egl_dmabuf.cc(608)] Failed to record frame: Error creating EGLImage - EGL_NOT_INITIALIZED
[181198:0312/033104.736966:ERROR:shared_screencast_stream.cc(940)] Dropping DMA-BUF modifier: 144115188757438212 and trying to renegotiate stream parameters
[181198:0312/033104.745879:ERROR:egl_dmabuf.cc(608)] Failed to record frame: Error creating EGLImage - EGL_NOT_INITIALIZED
[181198:0312/033104.745894:ERROR:shared_screencast_stream.cc(940)] Dropping DMA-BUF modifier: 144115188757658372 and trying to renegotiate stream parameters
[181198:0312/033104.762079:ERROR:egl_dmabuf.cc(608)] Failed to record frame: Error creating EGLImage - EGL_NOT_INITIALIZED
[181198:0312/033104.762090:ERROR:shared_screencast_stream.cc(940)] Dropping DMA-BUF modifier: 144115188757437188 and trying to renegotiate stream parameters
[181198:0312/033104.770421:ERROR:egl_dmabuf.cc(608)] Failed to record frame: Error creating EGLImage - EGL_NOT_INITIALIZED
[181198:0312/033104.770435:ERROR:shared_screencast_stream.cc(940)] Dropping DMA-BUF modifier: 144115188075858436 and trying to renegotiate stream parameters
[181198:0312/033104.787159:ERROR:egl_dmabuf.cc(608)] Failed to record frame: Error creating EGLImage - EGL_NOT_INITIALIZED
[181198:0312/033104.787168:ERROR:shared_screencast_stream.cc(940)] Dropping DMA-BUF modifier: 0 and trying to renegotiate stream parameters
[181198:0312/033104.795468:ERROR:egl_dmabuf.cc(608)] Failed to record frame: Error creating EGLImage - EGL_NOT_INITIALIZED
[181198:0312/033104.795482:ERROR:shared_screencast_stream.cc(940)] Dropping DMA-BUF modifier: 72057594037927935 and trying to renegotiate stream parameters
'loop->recurse > 0' failed at ../pipewire/src/pipewire/thread-loop.c:426 pw_thread_loop_wait()
```

</details>
<details> <summary>System/Version info</summary>

**Vesktop Information:**
- Stable 273850 (477ea08)
- Vencord 6e36381
- (Vesktop V1.5.1)
- Electron 29.1.1
- Chromium 122.9.6261.111

**Relevant System Information:**
- OS: Arch Linux
- Kernel: 6.7.9-arch1-1
- Hyprland Version: v0.36.0-100-ga065b481
- GPU: AMD Radeon RX 7900 XTX (radeonsi, navi31, LLVM 17.0.6, DRM 3.57, 6.7.9-arch1-1)
- Mesa Driver: 4.6 Mesa 24.0.2-arch1.2
- Display(s): 
   - [Samsung Electric Company Odyssey G7 H1AK500000](https://www.samsung.com/uk/monitors/gaming/odyssey-neo-g7-g70nc-43-inch-144hz-1ms-4k-uhd-ls43cg700nuxxu/)
     - 3840x2160@120 VRR 10bpc | modeline 1498.25 3840 4192 4616 5392 2160 2163 2168 2316 -HSync +VSync
   - [Ancor Communications Inc VG248 DCLMQS078349](https://www.asus.com/uk/displays-desktops/monitors/gaming/vg248qe/)
     - 1920x1080@60
   - [Dell Inc. DELL P2418D MY3ND7B70VLT](https://www.displayspecifications.com/en/model/9ee7d60)
     - 1920x1080@60

</details>

**Comment by RinLovesYou** [comment]
Trying this on nvidia is an exercise in futility, as the [community made vaapi driver](https://github.com/elFarto/nvidia-vaapi-driver) only supports decode.

**Comment by Vendicated** [comment]
> Running Vencord with:

by editing the source code, not via command line flags. scroll up ^

**Comment by rurigk** [comment]
I tested this but it has 2 problems

### First problem
Its not encoding with GPU because its missing the necessary flags https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/gpu/vaapi.md#verify-driver

If i add this to `index.ts` it starts to encode with GPU and its much smoother but not as smooth as it should be because the second problem
```
app.commandLine.appendSwitch("use-gl", "angle");
app.commandLine.appendSwitch("use-angle", "gl");
app.commandLine.appendSwitch("ozone-platform", "x11");
app.commandLine.appendSwitch("enable-features", "VaapiVideoDecodeLinuxGL,VaapiVideoEncoder,VaapiVP8Encoder,VaapiVP9Encoder,VaapiAV1Encoder");
```

### Second problem
The used resolution is not the selected resolution

Even if i select 480 15fps the real resolution used is 1440 165hz? (1440 is my monitor resolution and 165 its the refresh rate of my monitor but im not sure its using 165hz or something else) the thing is that it ends up using the source real size and not the selected one

This is and the used bitrate is probably the cause of not as good fps even when hardware encoding is used

#### Third problem i noticed but not really related, when audio is captured it makes my whole sistem audio crackle probably because the capture sample rate used is different to my system sample rate which is 48 kHz, no other capture software like OBS makes my system audio crackle

**Comment by kaitlynkittyy** [comment]
> I tested this but it has 2 problems
> ### First problem
> 
> Its not encoding with GPU because its missing the necessary flags https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/gpu/vaapi.md#verify-driver
> 
> If i add this to `index.ts` it starts to encode with GPU and its much smoother but not as smooth as it should be because the second problem
> 
> ```
> app.commandLine.appendSwitch("use-gl", "angle");
> app.commandLine.appendSwitch("use-angle", "gl");
> app.commandLine.appendSwitch("ozone-platform", "x11");
> app.commandLine.appendSwitch("enable-features", "VaapiVideoDecodeLinuxGL,VaapiVideoEncoder,VaapiVP8Encoder,VaapiVP9Encoder,VaapiAV1Encoder");
> ```
> 
> ### Second problem
> 
> The used resolution is not the selected resolution
> 
> Even if i select 480 15fps the real resolution used is 1440 165hz? (1440 is my monitor resolution and 165 its the refresh rate of my monitor but im not sure its using 165hz or something else) the thing is that it ends up using the source real size and not the selected one
> 
> This is and the used bitrate is probably the cause of not as good fps even when hardware encoding is used
> #### Third problem
> 
> Another thing i noticed but not really related, when audio is captured it makes my whole sistem audio crackle probably because the capture sample rate used is different to my system sample rate which is 48 kHz, no other capture software like OBS makes my system audio crackle

seems to be somewhat of an improvement for me too with these, but still not perfect, and from the people I asked, there are still spikes here and there.

System Specs:
Arch Linux: 6.8.1-zen
KDE Plasma 6 on Wayland
Ryzen 5 3600
RX 6800XT
32GB 3000MHz DDR4

**Comment by IlChitarrista** [comment]
> Might be a long-standing [chromium bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016) that we could work around

Okay, I might have something interesting to note about this.
I've found a way to push the framerate up from ~5 to ~25 and this workaround would explain what @MiMillieuh was initially seeing, i.e., focus playing some sort of role.
This is unrelated to hardware acceleration so if the maintainer agrees we could move this to a new issue.

When screencasting, the framerate, as reported by others, seems to dip when the frame-to-frame difference increases.
I've found that, if a video of any kind, stream or camera, is being run by Vesktop while streaming it will cause a significant increase in the quality received on the other end.

To reproduce this you can stream and test something intensive and ask someone on the other end to activate a camera (or use yours although this seems to have a slight performance hit on my end)

Here's a comparison, recorded from a different computer, stream is on the left while virtual obs camera is on the right.
In the first example the computer that streams has their own camera disabled while in the second one it isn't causing an immediate improvement in quality.

https://github.com/Vencord/Vesktop/assets/86124987/94078ca1-54d6-4f18-b6d4-213dfb2be1df

https://github.com/Vencord/Vesktop/assets/86124987/863610f8-00ba-4a7e-9131-3eb838b8cd53

I wonder if this might be preventing the aforementioned [chromium bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016)...

**Comment by kaitlynkittyy** [comment]
> > Might be a long-standing [chromium bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016) that we could work around
> 
> Okay, I might have something interesting to note about this. I've found a way to push the framerate up from ~5 to ~25 and this workaround would explain what @MiMillieuh was initially seeing, i.e., focus playing some sort of role. This is unrelated to hardware acceleration so if the maintainer agrees we could move this to a new issue.
> 
> When screencasting, the framerate, as reported by others, seems to dip when the frame-to-frame difference increases. I've found that, if a video of any kind, stream or camera, is being run by Vesktop while streaming it will cause a significant increase in the quality received on the other end.
> 
> To reproduce this you can stream and test something intensive and ask someone on the other end to activate a camera (or use yours although this seems to have a slight performance hit on my end)
> 
> Here's a comparison, recorded from a different computer, stream is on the left while virtual obs camera is on the right. In the first example the computer that streams has their own camera disabled while in the second one it isn't causing an immediate improvement in quality.
> no_camera.mp4
> camera.mp4
> 
> I wonder if this might be preventing the aforementioned [chromium bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016)...
This is quite interesting, I've been testing/messing with the experimental hardware acceleration flags, and I was also able to get it to about 20fps. 
Did you do these tests with the aforementioned flags, or were these just with stock vesktop? I'd love to see progress being made on this issue.

**Comment by kaitlynkittyy** [comment]
One more thing to add, as this issue is present in chromium itself, so a proper solution / issue will have to be forwarded to the respective maintainers. If a solution isn't found here, I'm hoping that someone there can find a solution using [this](https://issues.chromium.org/u/0/issues/331987171) issue.

**Comment by IlChitarrista** [comment]
> > > Might be a long-standing [chromium bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016) that we could work around
> > 
> > 
> > Okay, I might have something interesting to note about this. I've found a way to push the framerate up from ~5 to ~25 and this workaround would explain what @MiMillieuh was initially seeing, i.e., focus playing some sort of role. This is unrelated to hardware acceleration so if the maintainer agrees we could move this to a new issue.
> > When screencasting, the framerate, as reported by others, seems to dip when the frame-to-frame difference increases. I've found that, if a video of any kind, stream or camera, is being run by Vesktop while streaming it will cause a significant increase in the quality received on the other end.
> > To reproduce this you can stream and test something intensive and ask someone on the other end to activate a camera (or use yours although this seems to have a slight performance hit on my end)
> > Here's a comparison, recorded from a different computer, stream is on the left while virtual obs camera is on the right. In the first example the computer that streams has their own camera disabled while in the second one it isn't causing an immediate improvement in quality.
> > no_camera.mp4
> > camera.mp4
> > I wonder if this might be preventing the aforementioned [chromium bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016)...
> 
> This is quite interesting, I've been testing/messing with the experimental hardware acceleration flags, and I was also able to get it to about 20fps. Did you do these tests with the aforementioned flags, or were these just with stock vesktop? I'd love to see progress being made on this issue.

I'm running the Flatpak version of Vesktop on Gnome 46 (Fedora Rawhide/40) with an Nvidia GPU (550.67) with the following flags:

```flatpak run dev.vencord.Vesktop --enable-features=UseOzonePlatform --ozone-platform=wayland --use-gl=angle --use-angle=gl --enable-features=VaapiVideoDecodeLinuxGL,VaapiVideoEncoder,VaapiVP8Encoder,VaapiVP9Encoder,VaapiAV1Encoder,VaapiIgnoreDriverChecks,VaapiVideoDecoder,CanvasOopRasterization,UseMultiPlaneFormatForHardwareVideo```

This shouldn't really have a significant impact as Nvidia VA-API only implements decoders.
I have yet to test without the flags but, as for my previously quite unexplainable testing, people reported wide variability in quality which can now be linked to a camera keeping Vesktop active.

I've been able to reproduce this on stock [WebCord](https://github.com/SpacingBat3/WebCord) and I will soon get someone to test it on KDE as well.
The receiving client was on KDE (6?) and was tested with both Vesktop and Firefox.

I'm testing out a patched version of Pipewire which backports explicit modifiers support to the stable release and the stream comes from a Rawhide system which has GStreamer 1.24 but the findings are in line with my previous experience on 39/40.

The camera is still significantly smoother so I wonder about what could be done to better make use of the available APIs.
I.e., how can a virtual obs camera using Pipewire Screen Capture produce a significantly better image on the receiving end compared to the Vesktop's direct usage?
Does it have to do with how the camera stream is being encoded/compressed differently from the stream itself? Do you have any suggestions for potential further testing?

**Comment by kaitlynkittyy** [comment]
> > > > Might be a long-standing [chromium bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016) that we could work around
> > > 
> > > 
> > > Okay, I might have something interesting to note about this. I've found a way to push the framerate up from ~5 to ~25 and this workaround would explain what @MiMillieuh was initially seeing, i.e., focus playing some sort of role. This is unrelated to hardware acceleration so if the maintainer agrees we could move this to a new issue.
> > > When screencasting, the framerate, as reported by others, seems to dip when the frame-to-frame difference increases. I've found that, if a video of any kind, stream or camera, is being run by Vesktop while streaming it will cause a significant increase in the quality received on the other end.
> > > To reproduce this you can stream and test something intensive and ask someone on the other end to activate a camera (or use yours although this seems to have a slight performance hit on my end)
> > > Here's a comparison, recorded from a different computer, stream is on the left while virtual obs camera is on the right. In the first example the computer that streams has their own camera disabled while in the second one it isn't causing an immediate improvement in quality.
> > > no_camera.mp4
> > > camera.mp4
> > > I wonder if this might be preventing the aforementioned [chromium bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016)...
> > 
> > 
> > This is quite interesting, I've been testing/messing with the experimental hardware acceleration flags, and I was also able to get it to about 20fps. Did you do these tests with the aforementioned flags, or were these just with stock vesktop? I'd love to see progress being made on this issue.
> 
> I'm running the Flatpak version of Vesktop on Gnome 46 (Fedora Rawhide/40) with an Nvidia GPU (550.67) with the following flags:
> 
> `flatpak run dev.vencord.Vesktop --enable-features=UseOzonePlatform --ozone-platform=wayland --use-gl=angle --use-angle=gl --enable-features=VaapiVideoDecodeLinuxGL,VaapiVideoEncoder,VaapiVP8Encoder,VaapiVP9Encoder,VaapiAV1Encoder,VaapiIgnoreDriverChecks,VaapiVideoDecoder,CanvasOopRasterization,UseMultiPlaneFormatForHardwareVideo`
> 
> This shouldn't really have a significant impact as Nvidia VA-API only implements decoders. I have yet to test without the flags but, as for my previously quite unexplainable testing, people reported wide variability in quality which can now be linked to a camera keeping Vesktop active.
> 
> I've been able to reproduce this on stock [WebCord](https://github.com/SpacingBat3/WebCord) and I will soon get someone to test it on KDE as well. The receiving client was on KDE (6?) and was tested with both Vesktop and Firefox.
> 
> I'm testing out a patched version of Pipewire which backports explicit modifiers support to the stable release and the stream comes from a Rawhide system which has GStreamer 1.24 but the findings are in line with my previous experience on 39/40.
> 
> The camera is still significantly smoother so I wonder about what could be done to better make use of the available APIs. I.e., how can a virtual obs camera using Pipewire Screen Capture produce a significantly better image on the receiving end compared to the Vesktop's direct usage? Does it have to do with how the camera stream is being encoded/compressed differently from the stream itself? Do you have any suggestions for potential further testing?


I believe (I'm no expert of course) that the issue comes from the way that chromium itself handles screencapture, which then perpetuates the issue into electron, and finally vesktop. I'm sort of at a loss for future testing, and I'm just waiting on the chromium team to get back to me in the issue mentioned above. Also if you need/want me to test anything that you find out, lmk.

**Comment by IlChitarrista** [comment]
> > > > > Might be a long-standing [chromium bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016) that we could work around
> > > > 
> > > > 
> > > > Okay, I might have something interesting to note about this. I've found a way to push the framerate up from ~5 to ~25 and this workaround would explain what @MiMillieuh was initially seeing, i.e., focus playing some sort of role. This is unrelated to hardware acceleration so if the maintainer agrees we could move this to a new issue.
> > > > When screencasting, the framerate, as reported by others, seems to dip when the frame-to-frame difference increases. I've found that, if a video of any kind, stream or camera, is being run by Vesktop while streaming it will cause a significant increase in the quality received on the other end.
> > > > To reproduce this you can stream and test something intensive and ask someone on the other end to activate a camera (or use yours although this seems to have a slight performance hit on my end)
> > > > Here's a comparison, recorded from a different computer, stream is on the left while virtual obs camera is on the right. In the first example the computer that streams has their own camera disabled while in the second one it isn't causing an immediate improvement in quality.
> > > > no_camera.mp4
> > > > camera.mp4
> > > > I wonder if this might be preventing the aforementioned [chromium bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016)...
> > > 
> > > 
> > > This is quite interesting, I've been testing/messing with the experimental hardware acceleration flags, and I was also able to get it to about 20fps. Did you do these tests with the aforementioned flags, or were these just with stock vesktop? I'd love to see progress being made on this issue.
> > 
> > 
> > I'm running the Flatpak version of Vesktop on Gnome 46 (Fedora Rawhide/40) with an Nvidia GPU (550.67) with the following flags:
> > `flatpak run dev.vencord.Vesktop --enable-features=UseOzonePlatform --ozone-platform=wayland --use-gl=angle --use-angle=gl --enable-features=VaapiVideoDecodeLinuxGL,VaapiVideoEncoder,VaapiVP8Encoder,VaapiVP9Encoder,VaapiAV1Encoder,VaapiIgnoreDriverChecks,VaapiVideoDecoder,CanvasOopRasterization,UseMultiPlaneFormatForHardwareVideo`
> > This shouldn't really have a significant impact as Nvidia VA-API only implements decoders. I have yet to test without the flags but, as for my previously quite unexplainable testing, people reported wide variability in quality which can now be linked to a camera keeping Vesktop active.
> > I've been able to reproduce this on stock [WebCord](https://github.com/SpacingBat3/WebCord) and I will soon get someone to test it on KDE as well. The receiving client was on KDE (6?) and was tested with both Vesktop and Firefox.
> > I'm testing out a patched version of Pipewire which backports explicit modifiers support to the stable release and the stream comes from a Rawhide system which has GStreamer 1.24 but the findings are in line with my previous experience on 39/40.
> > The camera is still significantly smoother so I wonder about what could be done to better make use of the available APIs. I.e., how can a virtual obs camera using Pipewire Screen Capture produce a significantly better image on the receiving end compared to the Vesktop's direct usage? Does it have to do with how the camera stream is being encoded/compressed differently from the stream itself? Do you have any suggestions for potential further testing?
> 
> I believe (I'm no expert of course) that the issue comes from the way that chromium itself handles screencapture, which then perpetuates the issue into electron, and finally vesktop. I'm sort of at a loss for future testing, and I'm just waiting on the chromium team to get back to me in the issue mentioned above. Also if you need/want me to test anything that you find out, lmk.

Sure, you could test this workaround for now if you want.

I'm also wondering whether https://github.com/Vencord/Vesktop/issues/463 plays some role here.
Theoretically the same video stream at 1440p would require higher bitrate to maintain a similar perceived quality relative to resolution compared to the 720p being expected in most cases.
It is true that the framerate/quality tradeoff is a necessary evil in this case but, as seen, Vesktop seems to prefer a higher quality at a lower framerate, which could be improved with more efficient codecs being used but would similarly be aided by a lower amount of data being sent overall through proper resolution selection.

I'm also under the impression that the the framerate selector isn't working properly either although I can't quite see how that would affect this and don't have any actual data on it yet.

**Comment by kaitlynkittyy** [comment]
From the people i tested with, they didn't perceive much of a difference with a video playing or not. I didn't collect any specific numbers but I was told that it looked to be around 10-15fps no matter what.

**Comment by IlChitarrista** [comment]
> From the people i tested with, they didn't perceive much of a difference with a video playing or not. I didn't collect any specific numbers but I was told that it looked to be around 10-15fps no matter what.

Interesting, what hardware are you using?
I've also noticed that when there isn't much of a difference per-frame the stream will be decently smooth already; this only seems to help on games with a lot of per-frame differences.

This is a simpler example with no camera, then camera. The reference obs camera is on the right in the first video and at the bottom for the second:

https://github.com/Vencord/Vesktop/assets/86124987/25fd1a34-7030-45dd-80dd-8f64ecb62265

https://github.com/Vencord/Vesktop/assets/86124987/a00f8265-38d6-44a4-8629-ec3c7a4563c4

**Comment by IlChitarrista** [comment]
I have some new findings; I've done some tests while in debug mode for the screenshare.

My current theory is that the fundamental issue that drops the framerate is hitting the bitrate cap; this is presented as two independent components:
- max bitrate is set at 2.500kbps
- resolution setting is broken (#463) sending more data than required

So, first of all, in Vesktop's [source code](https://github.com/Vencord/Vesktop/blob/4abae9c7082081dcae667916d9608e23adf688a9/src/renderer/components/ScreenSharePicker.tsx#L74C22-L74C23), the max bitrate is set at 80.000kbps while minimum is at 5.000kbps and, most importantly, target is at 6.000kbps
This is not respected by Chromium which has a hardcoded limit of 2.500kbps; there is, however, an [API](https://developer.mozilla.org/en-US/docs/Web/API/RTCRtpSender/setParameters#maxbitrate) through which to set a custom limit; I've found this through the following [thread](https://groups.google.com/g/discuss-webrtc/c/0CnZzG9y9nM) leading to [this](https://groups.google.com/g/discuss-webrtc/c/ORJdeoFAaBE).
We should test out a higher limit through the usage of this API.

The second issue, broken resolution setting, makes it so that the stream is always at the streamed monitor's resolution, as verifiable through the debug mode; this, in turn, makes us send more data overall then needed in most cases as Discord's free limit is 720p.
This also has negative performance implications, especially if the stream fps is similarly locked to the monitor's refresh rate.

I believe that these two issues can catastrophically sum together leading to less than 5fps streams. 
Here's some rough data that I've collected so far by monitoring the debug while streaming a fast moving video:
- 1440p: ~6fps
- 1080p: ~14fps
- 720p: ~30fps 

In all three cases the max bitrate target was hit.

A reproducible way to spike up the framerate temporarily is to move the mouse around: the stream should approximate 30fps breaking through the 2.500kbps barrier briefly (~4.000kbps in the tested case) before being throttled back down.
Interestingly enough, the major difference between Vesktop's stream and what's reported on Windows (I haven't used it in a long), is that, instead of throttling quality down when reaching this limit, the framerate drops instead.
This partly explains why slowly updating content is fine, since it requires lower data it doesn't hit the limit avoiding the throttle.

This would explain why the camera was always smoother while, nevertheless, using the same h264(102) encoder; it was simply sending less data by actually being at 720p30.

I still don't have any explanation or data as to why having a camera in preview slightly improves the quality but, by judging the new findings, perhaps, it's best to investigate these two points, especially the max bitrate; correctness first, performance later.

**Comment by kaitlynkittyy** [comment]
I do believe the root cause may be StreamSettings failing to apply. Not only would that remove the set bitrate of 80kbps, but it would also cause the resolution setting to fail. I'll test with that and see if I can find any improvements to stream quality/framerate.

**Comment by moritzuehling** [comment]
I'm new to the project, and would love to try to set the [`contentHint`](https://developer.mozilla.org/en-US/docs/Web/API/MediaStreamTrack/contentHint#value) to `"motion"` for the mediaStreamTrack in addition to the stream settings, but I haven't found a function in vesktop / vencord. 

Is there a way to set those?

**Comment by kaitlynkittyy** [comment]
> I'm new to the project, and would love to try to set the [`contentHint`](https://developer.mozilla.org/en-US/docs/Web/API/MediaStreamTrack/contentHint#value) to `"motion"` for the mediaStreamTrack in addition to the stream settings, but I haven't found a function in vesktop / vencord. This would favour available bitrate to be used for new frames instead of quality on existing ones.
> 
> Is there a way to set those?


I've been experimenting with setting not just contenthint, but fixing the entire API to properly raise the bitrate from 2.5kbps to something normal (80) which should in theory fix most of the screensharing issues. On that note I've run out of ideas of how to implement that, so any help would be appreciated lol. contentHint could *possibly* also be set if the API rewrite was successful, but I can make no promises.

**Comment by moritzuehling** [comment]
In theory, it should be possible to write a demo page to try out options quickly, using the plain chromium screenshare interface. Just something that encodes a media-track, and immediately decodes it.

Maybe I can try this out, to see how to make it slow / fast. 

Might help in understanding the right options to set for streams, and see what's possible inside chromium

**Comment by kaitlynkittyy** [comment]
That may be a good idea, but I do not believe its the encoding/decoding causing the problem. It may be a good idea to create a small WebRTC streaming application to try and test.
It seems to be the bitrate that actually causes the issue.

**Comment by kaitlynkittyy** [comment]
> I'm new to the project, and would love to try to set the [`contentHint`](https://developer.mozilla.org/en-US/docs/Web/API/MediaStreamTrack/contentHint#value) to `"motion"` for the mediaStreamTrack in addition to the stream settings, but I haven't found a function in vesktop / vencord. This would favour available bitrate to be used for new frames instead of quality on existing ones.
> 
> Is there a way to set those?

Oh and I didn't fully answer this, I understand what you mean, but I believe the issue is that the available bitrate is so incredibly low, that not much improvement can even be made with how low it is. Without fixing that, I doubt we're getting anywhere. Do feel free to try though, you may find something we did not!

**Comment by SpookySkeletons** [comment]
Where in the chromium codebase is the integer to adjust?
I can hard patch it to check

**Comment by Vendicated** [comment]
i recommend moving this discussion to the vesktop development channel on [our discord server](https://vencord.dev/discord)

**Comment by EmptyVisual** [comment]
> i recommend moving this discussion to the vesktop development channel on [our discord server](https://vencord.dev/discord)

How does one actually get access to talk in the dev channels? I only see core and plugin dev and both say I dont have permission to send messages in channel. There does not appear to be a role that's self assignable for developers/contributors and if we have to jump through several hoops just to discuss on discord it might just be easier to discuss our findings on this issue?

**Comment by kaitlynkittyy** [comment]
to all of you who have been annoyed about this problem, here.
It's solved in this PR.
https://github.com/Vencord/Vesktop/pull/489

**Comment by Curve** [comment]
Closed by #489.

**Comment by sim590** [comment]
Is it really fixed? My friends say it's always a slideshow when I stream. I select the default settings:

![image](https://github.com/Vencord/Vesktop/assets/4147254/37eef6e9-add3-472d-902a-56e878355af7)

Here's the version of the package I installed on Archlinux:

```
Name            : vesktop-bin
Version         : 1.5.2-2
Description     : A cross platform electron-based desktop app aiming to give you a snappier Discord experience with Vencord pre-installed
Architecture    : x86_64
URL             : https://github.com/Vencord/Vesktop
Licenses        : GPL-3.0-only
Groups          : None
Provides        : vesktop=1.5.2  vencord-desktop=1.5.2
Depends On      : electron29
Optional Deps   : None
Required By     : None
Optional For    : None
Conflicts With  : vesktop  vencord-desktop
Replaces        : None
Installed Size  : 16.13 MiB
Packager        : Unknown Packager
Build Date      : Mon 20 May 2024 07:51:09 AM
Install Date    : Mon 20 May 2024 07:51:23 AM
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : None
```

**Comment by polluxau** [comment]
> Is it really fixed? My friends say it's always a slideshow when I stream. I select the default settings:
> 
> ![image](https://github.com/Vencord/Vesktop/assets/4147254/37eef6e9-add3-472d-902a-56e878355af7)
> 
> Here's the version of the package I installed on Archlinux:
> 
> ```
> Name            : vesktop-bin
> Version         : 1.5.2-2
> Description     : A cross platform electron-based desktop app aiming to give you a snappier Discord experience with Vencord pre-installed
> Architecture    : x86_64
> URL             : https://github.com/Vencord/Vesktop
> Licenses        : GPL-3.0-only
> Groups          : None
> Provides        : vesktop=1.5.2  vencord-desktop=1.5.2
> Depends On      : electron29
> Optional Deps   : None
> Required By     : None
> Optional For    : None
> Conflicts With  : vesktop  vencord-desktop
> Replaces        : None
> Installed Size  : 16.13 MiB
> Packager        : Unknown Packager
> Build Date      : Mon 20 May 2024 07:51:09 AM
> Install Date    : Mon 20 May 2024 07:51:23 AM
> Install Reason  : Explicitly installed
> Install Script  : No
> Validated By    : None
> ```

Are you using nvidia?

**Comment by shenk-i** [comment]
Same thing happens to me on AMD, my friends say its a slideshow.

GPU: AMD RX 7800 XT
CachyOS

**Comment by sim590** [comment]
Sorry, I meant to respond before, but I forgot. No, I am not on Nvidia. I run an AMD rx 5700XT. The version of my MESA driver is 1:24.0.7-3 (archlinux).

**Comment by NoelleTGS** [comment]
Had the same issue for a couple months, every time I try and screenshare I need to stop it and start it with every different combination of settings until it randomly decides to work. Bitrate (Target) will frequently be stuck at double digits.

**Comment by kaitlynkittyy** [comment]
> Sorry, I meant to respond before, but I forgot. No, I am not on Nvidia. I run an AMD rx 5700XT. The version of my MESA driver is 1:24.0.7-3 (archlinux).

considering you're on arch, do you have the libva-mesa-driver installed?
it's required for VA-API hardware accelerated streaming :)

**Comment by kaitlynkittyy** [comment]
> Been having the same issue for a couple months, every time I try and screenshare I need to stop it and start it with every different combination of settings until it randomly decides to work. Bitrate (Target) will frequently be stuck at double digits.

interesting, do you mind sharing your system specs? (and if on AMD refer to my latest comment to see what may be a possible culprit)

**Comment by NoelleTGS** [comment]
> > Been having the same issue for a couple months, every time I try and screenshare I need to stop it and start it with every different combination of settings until it randomly decides to work. Bitrate (Target) will frequently be stuck at double digits.
> 
> interesting, do you mind sharing your system specs? (and if on AMD refer to my latest comment to see what may be a possible culprit)

Not on AMD unfortunately.

CPU: Ryzen 7 5700X
GPU: RTX 3060 Ti
Running EndeavourOS with Nvidia proprietary drivers, and running Vesktop through the Flatpak. Anything else in particular that'd be helpful?

**Comment by kaitlynkittyy** [comment]
> > > Been having the same issue for a couple months, every time I try and screenshare I need to stop it and start it with every different combination of settings until it randomly decides to work. Bitrate (Target) will frequently be stuck at double digits.
> > 
> > interesting, do you mind sharing your system specs? (and if on AMD refer to my latest comment to see what may be a possible culprit)
> 
> Not on AMD unfortunately.
> 
> CPU: Ryzen 7 5700X
> GPU: RTX 3060 Ti
> Running EndeavourOS with Nvidia proprietary drivers, and running Vesktop through the Flatpak. Anything else in particular that'd be helpful?

considering you're on Nvidia, the best thing we can do is hope they one day support VA-API, but aside from that, you're at the limits of your software encoder
try streaming a lower res + fps and see if the bitrate target is able to increase at all(it may not though, as a lower res and framerate wouldn't need as much bitrate anyways)

**Comment by NoelleTGS** [comment]
Gotcha. It's just confusing because like I said, I have to try over and over and EVENTUALLY it stabilizes, seemingly independent of the settings I pick.

**Comment by kaitlynkittyy** [comment]
> Gotcha. It's just confusing because like I said, I have to try over and over and EVENTUALLY it stabilizes, seemingly independent of the settings I pick.

that is a bit odd, but I may have an idea
try running the latest commit, as a few commits back, background throttling was disabled, and it helps with bitrate throttling in the background
it may be that yours is dropping to the background a lot
I highly doubt it, but it's worth a try, as that is a bit odd.

**Comment by shenk-i** [comment]
I do have to ask, are you guys using wayland or x11? because i'm a bit confused now

**Comment by polluxau** [comment]
> I do have to ask, are you guys using wayland or x11? because i'm a bit confused now 

We should be using wayland but some desktops are still on x11 or have experimental wayland support, shouldnt rlly matter tho

**Comment by sim590** [comment]
> > Sorry, I meant to respond before, but I forgot. No, I am not on Nvidia. I run an AMD rx 5700XT. The version of my MESA driver is 1:24.0.7-3 (archlinux).
> 
> considering you're on arch, do you have the libva-mesa-driver installed? it's required for VA-API hardware accelerated streaming :)

Yes, I have the package installed with version 1:24.0.7-3.

**Comment by regalialong** [comment]
FYI, it would be nice to keep the chatter to a minimum because this has a lot of participants that get CC'd for debugging which should be done in IM / on another issue.

For what it's worth, maybe the user should be warned in the streaming dialog that performance may be worse if Electron is going to use a software encoder (either because of VA-API encoding being unavailable or just using Nvidia)?

**Comment by Vendicated** [comment]
locking since the original issue has been resolved. for further discussion, either open a new issue or better yet join the discord community and talk about it in a thread in the vesktop channel
